### DISCUSSION TITLE
Show stopper question: to flip or not to flip

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#1013] [Announcements]
The current preview implementation lays out sibling nodes from top to down and from left to right.
It does not depend on the layout.

It has the following consequence: the layout changes as shown here

<img width="585" alt="up-down-map" src="https://user-images.githubusercontent.com/1558052/219865360-e87ae72a-b143-45ce-8082-2782352ec835.png">

when layout <img width="47" alt="l-down" src="https://user-images.githubusercontent.com/1558052/219865382-aba41985-2eec-4a57-a083-349eb9a8515f.png"> is replaced by <img width="44" alt="l-up" src="https://user-images.githubusercontent.com/1558052/219865393-0c0b8a68-6a7d-4f0b-9904-1696f7e4196a.png">

But is it actually how it should be?

The logic could be chosen in a different way so that both layouts mirror each other and nodes are placed from the bottom to the top when the second layout is selected.

The same considerations can also be relevant to horizontal layouts. The current implementation behaves like this:

<img width="867" alt="left-right-map" src="https://user-images.githubusercontent.com/1558052/219865516-a5f0cf8f-73cb-476c-b3c6-d52b8effc628.png">

Please write your suggestions.

**Comment by sifran-github** [comment]
Giving my thoughts on the issue, maybe it will help:

I think that different configurations should be configurable by the user, either by:

1. by configuring the direction of the child nodes separately (children horizontal direction, children vertical direction)
2. or by giving options of different layouts, for example:
![image](https://user-images.githubusercontent.com/5730355/219940406-1ca0eeb3-7267-473e-9170-151a8a79a1dc.png)
![image](https://user-images.githubusercontent.com/5730355/219940417-979fa11f-d61a-45ed-ad34-687de246a7eb.png)

but i think that there are already a lot of layouts, and maybe it is better to reduce the number of layouts by giving the user another option to choose from, instead of multiplying all the layouts for this option.

**Reply by dpolivaev** [reply]
If you think that the both options are valuable, it actually simplifies development because I can keep ![image](https://user-images.githubusercontent.com/5730355/219940417-979fa11f-d61a-45ed-ad34-687de246a7eb.png) now and develop ![image](https://user-images.githubusercontent.com/5730355/219940406-1ca0eeb3-7267-473e-9170-151a8a79a1dc.png)
later as a new option.

Do you think that the existing option makes sense?

**Reply by dpolivaev** [reply]
I would not add a new layout but rather handle this option separately, and it could be done either at the node level or at the map level. But don't expect it to come very soon.

**Reply by sifran-github** [reply]
I currently don't use bottom-to-top layouts at all (even in other mind map programs which support it), but I can understand people that will use both versions.... so its hard for me to choose which layout makes more sense.

In SimplMind I found this "Linear layout" which looks like this:
![image](https://user-images.githubusercontent.com/5730355/219961951-1686e513-ee2e-4c2e-90a1-c7df9deaf2cf.png)
So here, the nodes order goes from the bottom to the top (on the upper part of the mind map).

Again, i think the best way is to allow users to choose for themselves.

General note about the new layouts: I actually really like the fact that you can choose any node and change its layout. its smart. its giving a lot of control to the user.
In SimpleMind for example you can only define the root structure and that's it, and the entire structure is being created. but it is very limiting.
What I really like about Freeplane is the ability for the user to define and change everything + with templates, styles, scripts and more tools that are in FreePlane, specific configurations can be automated, or that they are just easy to use (once they are defined).

**Comment by macmarrum** [comment]
I often use the node alignments FIRST_CHILD_BY_PARENT and LAST_CHILD_BY_PARENT.

Here's how the current implementation looks like:

![LAST_CHILD_BY_PARENT_2023-02-19_18,11,41](https://user-images.githubusercontent.com/7197131/219963597-c07ac2bd-48f3-4803-935d-5cfaed59208f.png)

![FIRST_CHILD_BY_PARENT_2023-02-19_18,12,27](https://user-images.githubusercontent.com/7197131/219963603-b4bf790c-4bc1-49a2-845b-094d24c1f63e.png)

It might be useful to have an option to **flip** the order. I believe that it should be allowed to flip the order of nodes for any layout. I can imagine a button :arrows_clockwise: which works as in the below picture (flipped on the right)

![regular_vs_flipped_2023-02-19_18,52,00](https://user-images.githubusercontent.com/7197131/219966320-7ae3721e-a2d0-4f11-9b33-34a038019f44.png)

The way I understand it, this **flip** would be a view option only. This means that when enumerating the nodes using a Groovy script `node.children.collect { it.text }` on the root, I would always get `['1', '2', '3', '4']`, regardless of whether the view is flipped or not.

**Reply by dpolivaev** [reply]
Yes, thank you, it is exactly the kind of answer I was asking for.

**Reply by sifran-github** [reply]
@dpolivaev was this ever implemented somehow? will it be? I wonder since you also liked the answer, and i find it useful.

and i hope it wont create more bugs because it adds complexity :)

**Reply by dpolivaev** [reply]
I have not started implementing it because currently, Freeplane assumes that all child nodes are places with ascending y (or x) coordinates. So it is not a small change and I want to release current version and integrate ChatGPT before spending time on this issue.
